feat: unified CLI

aaronstillwell opened the pull request:
This PR moves the project away from standalone python scripts and into a unified CLI.

This is intended on improving usability for new users and reducing time to value. It does so by providing a common set of arguments for each function of the CLI, as well as allowing the CLI to be used directly from the docker container or via a globally installed package.

## Before

Before this change, a user needs to checkout the repo, navigate to the correct directory, install dependencies, then run e.g

```
CIRCLECI_TOKEN=token ORG_ID=id START_DATE=2025-01-01 END_DATE=2025-02-01 python src/get_usage_report.py
```

To do so also may involve reading through the source code to understand which environment variables to specify.

## After

After this change, the user can install the CLI globally, or access it directly from the docker container, as well as get help on which arguments to specify.

For example:

```
pip install -e . # install system-wide
circleci-usage-reporter --help
circleci-usage-reporter get --org-id <id> --start-date 2025-01-01 --end-date 2025-02-01
```

Or, via docker:
```
docker build . -t circleci-usage-reporter:latest
docker run --rm circleci-usage-reporter:latest --help
```

# Breaking Changes

There are some additional breaking changes included as part of the move towards a unified CLI.

## get command (previously get_usage_report.py)
* Environment variables `START_DATE`,`END_DATE`,`OUTPUT` must be provided as CLI arguments instead

## create-graph command
* `csv_file` is now a required positional argument when using `create-graph`

## run-analysis command
* `--data-file` renamed to `--input`

Commit:
feat: unified CLI

This commit moves the project away from standalone python scripts and into a unified CLI

Commit:
fix: lint failures

Commit:
fix: non-existent package being used in pyproject.toml

Commit:
fix: docker container test

aaronstillwell commented:
Will effectively close #14

Commit:
chore: remove env vars for get_usage_report.py arguments

Commit:
fix: get_usage_report.py arg defs

Commit:
chore: move get_usage_report.py to get.py to align to cmd name

Commit:
docs: update examples

Commit:
chore: add new pip install to README.md

Commit:
chore: revert to previous get logic

jenny-miggin committed:
fix: fix run-analysis command bugs found during PR review testing

- Fix notebook executing in wrong cwd causing `import analysis` to fail:
  write temp notebook to src/ and pass cwd=src_dir to subprocess.run
- Fix --input flag being silently ignored: thread data_file through
  run_existing_notebook_with_params, create_minimal_notebook, and all
  four create_*_cells functions instead of hardcoding /tmp/merged.csv
- Fix relative input paths failing when notebook runs from src/:
  resolve args.data_file to absolute path in handle() before use
- Fix nbconvert running in wrong Python env: replace bare `jupyter`
  command with sys.executable -m nbconvert; add nbconvert and ipykernel
  to pyproject.toml dependencies
- Fix temp notebook files leaking on failure: raise RuntimeError instead
  of sys.exit in run_notebook_conversion so handle() try/finally block
  always cleans up the temp .ipynb file
- Add tmp*.ipynb to .gitignore to prevent future leaks being committed

Made-with: Cursor

jenny-miggin reviewed (commented):
## Review notes from hands-on testing

I tested this locally by installing via `pip install -e .` and running the CLI end-to-end against a real usage report. Several bugs in `run-analysis` needed fixing before it would work. Summarising findings below (with suggested fixes inline).

---

### Bugs found during testing

#### 1. `run-analysis`: `import analysis` fails — notebook runs in wrong directory

`run_notebook_conversion` calls `subprocess.run` without a `cwd` argument, so `nbconvert` executes the notebook from whatever directory the user ran the CLI from. The notebook does `import analysis`, which only resolves if the working directory is `src/`.

Also, the temp notebook is written to `dir="."` (user's CWD), but `nbconvert` is passed just the filename with no path, so it can't find the file either.

**Fix:**
- Write the temp notebook into `src/` using `Path(__file__).parent`
- Pass `cwd=str(Path(__file__).parent)` to `subprocess.run`

```python
# create_minimal_notebook
src_dir = Path(__file__).parent
temp_notebook = tempfile.NamedTemporaryFile(..., dir=str(src_dir))

# run_notebook_conversion
src_dir = str(Path(__file__).parent)
subprocess.run(cmd, ..., cwd=src_dir)
```

---

#### 2. `run-analysis`: `--input` flag is silently ignored

`args.data_file` is validated and used for `find_highest_credit_project`, but is never passed into `run_existing_notebook_with_params` or `create_minimal_notebook`. All four cell-generator functions hardcode `/tmp/merged.csv` in the generated notebook source, so `--input usage_report.csv` has no effect.

**Fix:** thread `data_file` through the call chain:

```python
# handle()
notebook_path = run_existing_notebook_with_params(
    ...,
    data_file=args.data_file
)

# run_existing_notebook_with_params signature
def run_existing_notebook_with_params(..., data_file='/tmp/merged.csv'):
    env_vars = {'FILEPATH': data_file, ...}
    ...
    return create_minimal_notebook(..., data_file)

# create_minimal_notebook + all four create_*_cells functions
def create_job_analysis_cells(project_name, individual_job_name, credit_cost, data_file='/tmp/merged.csv'):
    ...
    f"    filepath='{data_file}',\n",
```

---

#### 3. `run-analysis`: relative input paths break because notebook runs from `src/`

Even after fix #2, passing `--input usage_report.csv` (a relative path) still fails. Since the notebook kernel runs with `cwd=src/`, it looks for `src/usage_report.csv`.

**Fix:** resolve to absolute path immediately in `handle()`:

```python
def handle(args):
    args.data_file = str(Path(args.data_file).resolve())
    if not Path(args.data_file).exists():
        ...
```

---

#### 4. `pyproject.toml`: `nbconvert` and `ipykernel` missing from dependencies

`run-analysis` requires `nbconvert` and `ipykernel` to execute notebooks, but they aren't declared as dependencies. If a user installs with `pip install .` and their system Jupyter is in an isolated Homebrew/conda environment (common on macOS), the notebook executes in the wrong Python environment and `import pandas` fails.

**Fix:** add to `pyproject.toml` and use `sys.executable -m nbconvert` instead of bare `jupyter nbconvert`:

```toml
dependencies = [
    ...
    "nbconvert>=7.0.0",
    "ipykernel>=6.0.0"
]
```

```python
cmd = [sys.executable, "-m", "nbconvert", ...]
```

---

#### 5. Temp notebook files leak on failure

`handle()` cleans up the temp `.ipynb` on success (line 977), but `run_notebook_conversion` calls `sys.exit(1)` on error, which bypasses the cleanup. This leaves `tmp*.ipynb` files in `src/` on every failed run.

**Fix:** raise an exception instead of `sys.exit` in `run_notebook_conversion`, and wrap conversion in `try/finally` in `handle()`:

```python
# run_notebook_conversion — raise instead of sys.exit
except subprocess.CalledProcessError as e:
    raise RuntimeError(f"Notebook conversion failed: {e.stderr}") from e

# handle() — always clean up
try:
    output_file = run_notebook_conversion(...)
except RuntimeError as e:
    print(str(e), file=sys.stderr)
    return 1
finally:
    if Path(notebook_path).exists():
        os.unlink(notebook_path)
```

---

### Summary

The CLI itself (`get`, `merge`, `send-to-datadog`, `send-to-doit`, `create-graph`) all work correctly. The `run-analysis` command has 5 bugs that make it non-functional as-is — none of them are caught by the existing tests because there are no tests for the new code paths. Happy to discuss any of these or put up a follow-up PR with the fixes.

jenny-miggin committed:
fix: fix standalone main() entrypoints in all CLI modules

Each module's standalone main() was calling add_parser(parser) with a
plain ArgumentParser, but add_parser() expects a _SubParsersAction from
add_subparsers(). This caused AttributeError when running any module
directly (e.g. python src/merge.py --help).

Follows the pattern already used in get.py: extract _add_arguments()
as a shared helper that accepts any parser-like object, then call it
from both add_parser() (for the unified CLI) and main() (for standalone
execution).

Fixes merge.py, create_graph.py, run_analysis.py, and send_to_doit.py.

Made-with: Cursor

jenny-miggin committed:
chore: merge main into add-cli, resolve run_analysis.py conflict

Made-with: Cursor
